examples/energy-management-app/energy-management-common/src/ChargingTargetsMemMgr.cpp

```diff
@@ -0,0 +1,154 @@
+            chip::Platform::MemoryAlloc(sizeof(EnergyEvse::Structs::ChargingTargetStruct::Type) * mNumDailyChargingTargets));
+
+        for (uint16_t idx = 0; idx < mNumDailyChargingTargets; idx++)
+        {
+            // This will cause the ChargingTargetStruct constructor to be called and this element in the array
+            new (mpListOfDays[mChargingTargetSchedulesIdx] + idx) EnergyEvse::Structs::ChargingTargetStruct::Type();
+
+            // Now copy the chargingTarget
+            mpListOfDays[mChargingTargetSchedulesIdx][idx] = mDailyChargingTargets[idx];
+        }
+    }
+}
+
+void ChargingTargetsMemMgr::AllocAndCopy(const DataModel::List<const Structs::ChargingTargetStruct::Type> & chargingTargets)
+{
+    mNumDailyChargingTargets = static_cast<uint16_t>(chargingTargets.size());
+
+    if (mNumDailyChargingTargets > 0)
+    {
+        // Allocate the memory first and then use placement new to initialise the memory of each element in the array
+        mpListOfDays[mChargingTargetSchedulesIdx] = static_cast<EnergyEvse::Structs::ChargingTargetStruct::Type *>(
+            chip::Platform::MemoryAlloc(sizeof(EnergyEvse::Structs::ChargingTargetStruct::Type) * chargingTargets.size()));
+
+        uint16_t idx = 0;
+        for (auto & chargingTarget : chargingTargets)
+        {
+            // This will cause the ChargingTargetStruct constructor to be called and this element in the array
+            new (mpListOfDays[mChargingTargetSchedulesIdx] + idx) EnergyEvse::Structs::ChargingTargetStruct::Type();
+
+            // Now copy the chargingTarget
+            mpListOfDays[mChargingTargetSchedulesIdx][idx] = chargingTarget;
+
+            idx++;
+        }
+    }
+}
+
+CHIP_ERROR
+ChargingTargetsMemMgr::AllocAndCopy(const DataModel::DecodableList<Structs::ChargingTargetStruct::DecodableType> & chargingTargets)
+{
+    size_t numDailyChargingTargets = 0;
+    CHIP_ERROR err                 = chargingTargets.ComputeSize(&numDailyChargingTargets);
+    if (err == CHIP_NO_ERROR)
+    {
+        mNumDailyChargingTargets = static_cast<uint16_t>(numDailyChargingTargets);
+
+        if (mNumDailyChargingTargets > 0)
+        {
+            // Allocate the memory first and then use placement new to initialise the memory of each element in the array
+            mpListOfDays[mChargingTargetSchedulesIdx] = static_cast<EnergyEvse::Structs::ChargingTargetStruct::Type *>(
+                chip::Platform::MemoryAlloc(sizeof(EnergyEvse::Structs::ChargingTargetStruct::Type) * mNumDailyChargingTargets));
+
+            uint16_t idx = 0;
+            auto it      = chargingTargets.begin();
+            while (it.Next())
+            {
+                auto & chargingTarget = it.GetValue();
+
+                // This will cause the ChargingTargetStruct constructor to be called and this element in the array
+                new (mpListOfDays[mChargingTargetSchedulesIdx] + idx) EnergyEvse::Structs::ChargingTargetStruct::Type();
```

Comment:
how do we guarantee that idx does not iterate longer than `mNumDailyChazrgingTargets` and overflows?

Reply:
A check is made/error returned if we would exceed the `mNumDailyChargingTargets`.